### PR TITLE
Backend: Add ability to skip migration + use ability to prevent running migrations that break with MySQL

### PR DESCRIPTION
**Pull request by shayancanonical**
**What this PR does / why we need it**:
As noted in [this issue](https://github.com/grafana/grafana/issues/42177), Grafana backend cannot currently be used with MySQL with option `sql_require_primary_key=ON`. 

The reason being, the migrations create a table without a primary key, then rename said table, create new table with primary key, and then copy data from original table into new table. When MySQL has the aforementioned flag turn on, the creation of the original table fails.

As a solution, this PR presents a way to skip migrations if a condition is met. If the database does not already contain the original table (without the primary key), the migration that creates the original table will be skipped. This will result in the correct table with a primary key being created.

If the original table (without the primary key) already exists (for exisiting deployments of Grafana backend), then the new table with the primary key will be created and data copied from the original table into the new table.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/sqlstore/migrations/alert_mig.go [hunk 1]
```diff
@@ -55,13 +55,20 @@ func addAlertMigrations(mg *Migrator) {
 		},
 	}
 
-	mg.AddMigration("Create alert_rule_tag table v1", NewAddTableMigration(alertRuleTagTable))
-	mg.AddMigration("Add unique index alert_rule_tag.alert_id_tag_id", NewAddIndexMigration(alertRuleTagTable, alertRuleTagTable.Indices[0]))
+	alertRuleTableDoesNotExist := &IfNotTableExistsWithoutPrimaryKey{TableName: alertRuleTagTable.Name, ColumnName: "id"}
```
Comment (shayancanonical) [inline pkg/services/sqlstore/migrations/alert_mig.go:58]: rename `alertRuleTableDoesNotExist` -> `alertRuleTableDoesNotContainPrimaryKey`

File: pkg/services/sqlstore/migrations/annotation_mig.go [hunk 2]
```diff
@@ -79,13 +79,20 @@ func addAnnotationMig(mg *Migrator) {
 		},
 	}
 
-	mg.AddMigration("Create annotation_tag table v2", NewAddTableMigration(annotationTagTable))
-	mg.AddMigration("Add unique index annotation_tag.annotation_id_tag_id", NewAddIndexMigration(annotationTagTable, annotationTagTable.Indices[0]))
+	annotationTagTableDoesNotExist := &IfNotTableExistsWithoutPrimaryKey{TableName: annotationTagTable.Name, ColumnName: "id"}
```
Comment (shayancanonical) [inline pkg/services/sqlstore/migrations/annotation_mig.go:82]: rename `annotationTagTableDoesNotExist` -> `annotationTagTableDoesNotContainPrimaryKey`

File: pkg/services/sqlstore/migrations/alert_mig.go (hunk 1 above)
Comment (shayancanonical) [reply pkg/services/sqlstore/migrations/alert_mig.go]: renamed appropriately in 03e12ac7583fd2f78efcbf82d953c7fd46431d2e

File: pkg/services/sqlstore/migrations/annotation_mig.go (hunk 2 above)
Comment (shayancanonical) [reply pkg/services/sqlstore/migrations/annotation_mig.go]: renamed appropriately in 03e12ac7583fd2f78efcbf82d953c7fd46431d2e

File: pkg/services/sqlstore/migrator/migrator.go [hunk 3]
```diff
@@ -185,9 +185,28 @@ func (mg *Migrator) run() (err error) {
 	return mg.DBEngine.Sync2()
 }
 
-func (mg *Migrator) exec(m Migration, sess *xorm.Session) error {
+func (mg *Migrator) exec(m Migration, sess *xorm.Session) (bool, error) {
```
Comment (yuri-tceretian) [inline pkg/services/sqlstore/migrator/migrator.go:188]: The meaning of the result is not very clear. It would be great to have some docs for the method

**Comment by ying-jeanne** [comment]
Hey @shayancanonical, thanks for the contribution after several rounds of discussion we decided not to merge the PR. Here is a walkaround:
- disable temporally sql_require_primary_key and run the migration.
- create the table with the primary key and add the entry in the migration log.
We decided not to support the skip migration mainly because it could be misused, it would be a backdoor that we open for potential DB issues in the future, and for customer support, it would be much easier to debug issues since they would be much more difficult to know what DB migrations have been run previously.

**Comment by simskij** [comment]
Unfortunately, this is not possible with a  MySQL Group Replication cluster. Any other solution you would entertain? We are happy to contribute engineering time to make it happen.

**Comment by taurus-forever** [comment]
Dear @ying-jeanne ,

Thank you for the decision and your reply!

We are using MySQL Group Replication feature which has strict requirements on [sql_require_primary_key](https://dev.mysql.com/doc/refman/8.0/en/group-replication-requirements.html).
Handling Grafana DB updates in 3rd party components is a road to huge problems in the future.

Do you see another workarounds/solution here? At the moment it is a show stopper for us.
Thank you for understanding!

**Comment by iXingo** [comment]
same here, is there any solution for that

> Bumping this. @ying-jeanne, any suggestions on how to solve this on MySQL servers with group replication enabled? It is not really feasible to temporarily opt out of the feature on a live cluster.

**Pull request closed without merging**